### PR TITLE
feat(tooling): identify agent notification origins

### PR DESCRIPTION
**Pull request by kriegcloud**
Generic agent notifications do not identify the session that needs attention.
This change shows the agent, owning clone, and linked worktree on local desktop
notifications, and adds navigation where the originating session is known.

- Ghostty receives OSC 777 through a terminal descriptor retained before the
  hook detaches, allowing its native notification action to select the emitting
  terminal surface.
- Codex Desktop gets an Open task action when its task ID matches the hook's
  session ID. An inherited parent task ID cannot redirect a child CLI session.
- Launchers can provide an exact, validated Codex or Claude Desktop session URI.
  Arbitrary URLs, cross-agent routes, and last-session links are rejected.
- Clone names and navigation destinations stay out of telemetry and ntfy.
  Existing wait attribution and phone escalation remain. Desktop reminder stages
  are damped while the original actionable notification has a live listener.

T3 Code task routing and a Grok event adapter are not included. Claude Desktop
requires a launcher-provided local session URI; its ID is not inferred from the
CLI session UUID. Each wait keeps at most one live desktop action listener, with
a one-hour deadline. Clicks recheck that the original wait is still open.

Validation:

- 87 focused notifier, hook-writer, and circuit-breaker tests passed.
- Package verification (lint and source/test type checks) passed.
- Effect Vitest policy lint passed with zero introduced findings.
- Real PTY smoke verified detached Ghostty emission and silent hook streams.
- Codex Desktop smoke verified exact-ID routing, mismatched inherited IDs and
  explicit overrides, and suppression of clicks after the wait resolves.
- PTY regression covers all three Ghostty stages with authenticated ntfy; listener
  regression verifies only one live listener across reminder stages.
- Real Plasma/libnotify probe acknowledged the actionable notification; closing
  that probe did not open an application.
- Shell syntax and diff whitespace checks passed.

Actual interactive focus of each application's tab has not been manually
verified. Full local proof and hosted checks are tracked by Yeet.

---

## Provenance

- Workspace: <code>beep-effect19</code>
- Branch: <code>codex/notification-origins</code>
- Agents (newest first):
  - `codex` (tui) · `unknown` · monitored

Resume the publishing agent from any beep-effect checkout on the publishing workstation:

```sh
bun run beep yeet resume 1144
```

## Summary by CodeRabbit

- **New Features**
  - Improved desktop notifications with clearer agent, clone, and worktree context.
  - Added Ghostty terminal notifications when supported.
  - Added actionable notifications that can open validated Codex or Claude session links.
  - Added safeguards to prevent unsupported or unrelated links from being opened.
- **Documentation**
  - Documented notification behavior, supported integrations, privacy boundaries, and fallback behavior.

### CODE REVIEW

File: .claude/hooks/sequence-break-notifier.sh [hunk 1]
```diff
@@ -385,6 +391,70 @@ notification_body() {
...
+
+# A bounded background listener owns the notification action. Its stdout is
+# private to this pipe, never the permission hook. Receipt of the notification
+# ID proves the send; closing/dismissing never opens an app. Reminder delivery
+# does not wait on this listener. No session URL is written to the ledger.
+deliver_action_notification() {
+  local stage="${1}" measured_age="${2}" urgency="${3}" title="${4}" body="${5}"
+  (
+    XDG_RUNTIME_DIR="${runtime_dir}" DBUS_SESSION_BUS_ADDRESS="${bus_address}" \
+      timeout 3600s stdbuf -oL notify-send --app-name="beep agent" --urgency="${urgency}" --expire-time=0 \
```
[the remark this answers is not in the document]
Comment (kriegcloud) [reply .claude/hooks/sequence-break-notifier.sh:441]: Verified this against Bash: a pipeline subshell created inside a function retains the function context, so local is legal here. A reproduction using bash -euo pipefail with the same pipeline/read shape succeeds and prints 42:default. The notifier also sets -uo pipefail, not -e. The new action regression test passes, and a real Plasma/libnotify probe recorded desktop:sent before interaction; the stubbed click test opened exactly the allowlisted URI. No code change is needed for this finding.

File: .claude/hooks/sequence-break-notifier.sh [hunk 2]
```diff
@@ -385,6 +391,70 @@ notification_body() {
...
+        case "${notification_id}" in
+          "" | *[!0-9]*) append_delivery desktop "${stage}" failed command-failed "${measured_age}"; exit 0 ;;
+        esac
+        append_delivery desktop "${stage}" sent "" "${measured_age}"
+        if IFS= read -r action && [ "${action}" = "default" ] && [ ! -e "${disarm_sentinel}" ]; then
+          XDG_RUNTIME_DIR="${runtime_dir}" DBUS_SESSION_BUS_ADDRESS="${bus_address}" \
+            timeout 5s xdg-open "${open_uri}" </dev/null >/dev/null 2>&1 || true
+        fi
+      }
+  ) </dev/null >/dev/null 2>&1 &
```
Comment (kriegcloud) [inline .claude/hooks/sequence-break-notifier.sh:459]: **P2 · Each stage can spawn a 3600s action listener**

`deliver_action_notification` backgrounds a `timeout 3600s notify-send --print-id --action=…` listener and returns immediately. When `open_uri` is set, `deliver_stage` may invoke this for initial, reminder, and urgent — up to three hour-long listeners for one wait. Dismiss-without-open and the disarm check before `xdg-open` are correct; the concern is unbounded-ish concurrent listeners under multi-stage escalation (storm damping limits new waiters, not stages within one worker).

Prompt To Fix With AI:
In `deliver_action_notification` / `deliver_desktop`, prefer a single action listener per worker (replace or skip if one is already live) or shorten the action timeout to align with reminder cadence (e.g. slightly above `BEEP_SEQUENCE_BREAK_URGENT_SECONDS`). Document the max concurrent listeners in `docs/runbooks/agent-notifications.md`. Optional test: with `open_uri` set and `maxStage=urgent`, assert at most one `notify-send --action` child is active at a time (or that later stages reuse/replace).

File: packages/tooling/library/ai-metrics/test/sequence-break.test.ts [hunk 3]
```diff
@@ -183,11 +183,23 @@ const runNotifier = Effect.fnUntraced(function* (
...
+    [
+      "claude-code",
+      SESSION_ID,
+      REQUEST_TS,
+      "tool-permission",
+      "human-input",
+      "AskUserQuestion",
+      "desktop-ntfy-1",
+      originCwd,
+      openUri,
```
Comment (kriegcloud) [inline packages/tooling/library/ai-metrics/test/sequence-break.test.ts]: **P2 · runNotifier never passes origin_terminal (arg 10); Ghostty OSC untested**

The notifier contract adds args 8/9/10 (`origin_cwd`, `open_uri`, `origin_terminal`), but `runNotifier` only forwards `originCwd` and `openUri` — arg 10 is always empty in the suite. New tests cover labeling, worktree, action-open, and bad URIs, yet nothing exercises `origin_terminal=ghostty` or the OSC 777 / FD 8 path that the runbook treats as a primary return path. That also means the FD 8 × ntfy collision above would not have been caught here.

Prompt To Fix With AI:
Extend `runNotifier` with an `originTerminal` parameter (default `""`) passed as argv[10]. Add an effect test that sets `origin_terminal=ghostty`, leaves `open_uri` empty, provides a writable tty-like FD 8 (or spies the `bash -c printf` OSC helper), and asserts the Ghostty path is taken (no `--action` / no plain notify-send body path as appropriate). After the FD fix, also assert a follow-up stage still works when an ntfy token was used.

File: .claude/hooks/sequence-break-notifier.sh [hunk 4]
```diff
@@ -393,17 +463,46 @@ deliver_desktop() {
...
+  # write terminal escapes to hook stdout or reopen a potentially recycled PTY.
+  if [ -z "${open_uri}" ] && [ "${origin_terminal}" = "ghostty" ] && [ -t 8 ]; then
+    local osc_body
+    # Ghostty uses the body as its notification ID. Distinguish simultaneous
+    # sessions in the same checkout so one cannot replace the other's action.
+    osc_body="$(jq -nr --arg body "${body} [${session_id:0:12}]" '$body | gsub("[\u0000-\u001f\u007f-\u009f;]"; " ")')"
+    if [ ! -e "${disarm_sentinel}" ] &&
+      timeout 2s bash -c 'printf "\\033]777;notify;%s;%s\\033\\\\" "$1" "$2" >&8' bash "${title}" "${osc_body}"; then
+      append_delivery desktop "${stage}" sent "" "${measured_age}"
+      return 0
```
Comment (kriegcloud) [inline .claude/hooks/sequence-break-notifier.sh:503]: **P1 · Ghostty FD 8 collides with ntfy bearer path — reminder/urgent OSC broken when token set**

This new path requires a live tty on FD 8 (`[ -t 8 ]`, `printf … >&8`). Pre-existing `deliver_ntfy` (same file, token branch) does `exec 8< <(Authorization…)` then `exec 8<&-` after every tokenized send. `deliver_stage` always runs `deliver_desktop` then `deliver_ntfy`, so an initial Ghostty send succeeds and then ntfy closes FD 8 — reminder/urgent skip OSC and lose surface-bound click targeting exactly when the operator most needs the originating tab. (Static repro: open FD 8 → ntfy close sequence → `[ -t 8 ]` false.)

Prompt To Fix With AI:
Stop sharing FD 8 between Ghostty and ntfy. At worker start, if `[ -t 8 ]`, `exec 7>&8` and emit OSC 777 on FD 7; keep ntfy bearer on FD 8 — or move the bearer header to FD 7/10 and never close the Ghostty tty. Add a regression: `origin_terminal=ghostty`, empty `open_uri`, non-empty ntfy token, after one `deliver_ntfy` a later stage can still OSC.

**Review by kriegcloud — COMMENTED** [review]
L1 Greptile-format review at 07b5c1f4: P0:0 P1:1 P2:2 — Ghostty FD 8 × ntfy bearer collision; Ghostty path untested; multi-stage 3600s action listeners. Do not APPROVE (mergeable_state blocked).

**Comment by kriegcloud** [comment]
## Confidence Score: 4/5

L1 tooling/hooks review: URI allowlist and privacy look solid; one verified P1 Ghostty×ntfy FD collision should be fixed before relying on Ghostty reminders with a bearer token. Do not APPROVE while `mergeable_state` is blocked.

## Findings

1. **P1 · ntfy bearer path closes FD 8 and breaks Ghostty OSC on reminder/urgent** → `.claude/hooks/sequence-break-notifier.sh` (`exec 8<` / `exec 8<&-`)
2. **P2 · runNotifier never passes origin_terminal (arg 10); Ghostty OSC untested** → `sequence-break.test.ts`
3. **P2 · Each stage can spawn a 3600s action listener** → `deliver_action_notification`

<details>
<summary>Fix with agent prompt</summary>

### Issue 1
.claude/hooks/sequence-break-notifier.sh (deliver_ntfy token branch ~630-641 + Ghostty OSC ~490-498)

Stop sharing FD 8 between Ghostty and ntfy. Dup the retained tty at worker start (e.g. `exec 7>&8` when `-t 8`) and emit OSC 777 on the dup; or move the bearer `Authorization` header to FD 7/10 and never `exec 8<&-` the Ghostty descriptor. Add a regression: Ghostty path + ntfy token + later stage still has a live tty FD for OSC.

---

### Issue 2
packages/tooling/library/ai-metrics/test/sequence-break.test.ts (`runNotifier`)

Add `originTerminal` argv[10]. Test `origin_terminal=ghostty` with empty `open_uri` (spy OSC / assert Ghostty path). After Issue 1, assert a post-ntfy stage can still OSC.

---

### Issue 3
.claude/hooks/sequence-break-notifier.sh (`deliver_action_notification`)

Cap at one action listener per worker or shorten the 3600s timeout to the escalation window; document max concurrency in `docs/runbooks/agent-notifications.md`.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
</details>

<details open>
<summary><h3>Summary</h3></summary>

- What the PR does: local desktop notifications label agent/clone/worktree; Ghostty OSC 777 via retained FD 8; validated `codex://` / `claude://` **Open task** actions via `xdg-open`; keep origin/URI out of evidence and ntfy; runbook + notifier regression tests.
- Commits: `544cdb42` feat + `07b5c1f4` concurrent-session OSC body suffix.
- Verified seats (Domain / Reliability / Junior): allowlist + `CODEX_THREAD_ID == raw_session_id` gate + privacy hold; **P1** FD 8 destroyed by ntfy bearer path after first `deliver_stage`; **P2** Ghostty arg/path untested; **P2** up to three 3600s action listeners per wait.
- Notes (not Findings): Codex thin wrapper correctly `exec`s shared worker with `"$@"`; docs lexicon matches `Open task` / Ghostty precedence.
- Risk tier L1 · seats: Domain Modeler, Reliability Adversary, Junior Engineer (Category Theorist skipped — no Effect/Schema surface)
- Graft: deep_meaning_used:false (DankStation offline) · reviewed from full PR diff + GitHub file contents at `07b5c1f4`

</details>

<details open>
<summary><h3>Diagram</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
  Hook["hook-pulse PermissionRequest"] --> FD8["exec 8>/dev/tty if Ghostty"]
  Hook --> Args["cwd + open_uri + terminal args"]
  Args --> Worker["sequence-break-notifier"]
  Worker --> Allow{"agent_kind:open_uri allowlist"}
  Allow -->|reject| Label["labeled notify-send only"]
  Allow -->|ok| Action["3600s notify-send --action + xdg-open"]
  Worker --> Ghostty{"open_uri empty AND -t 8"}
  Ghostty -->|yes| OSC["OSC 777 on FD 8"]
  Ghostty -->|no| Label
  Worker --> Ntfy["deliver_ntfy"]
  Ntfy -->|"token: exec 8 then close"| Break["FD 8 gone P1"]
  Break --> Reminder["reminder/urgent Ghostty skipped"]
```

</details>

<sub>Reviews (1) · Last reviewed commit: ["fix(tooling): distinguish concurrent notification sessions"](https://github.com/beep-effect/beep-effect/commit/07b5c1f401c26d19fd35ccb1a2dae2c44db87cbf)</sub>

File: .claude/hooks/sequence-break-notifier.sh [hunk 5]
```diff
@@ -385,6 +391,70 @@ notification_body() {
...
+          append_delivery desktop "${stage}" failed command-failed "${measured_age}"
+          exit 0
+        fi
+        case "${notification_id}" in
+          "" | *[!0-9]*) append_delivery desktop "${stage}" failed command-failed "${measured_age}"; exit 0 ;;
+        esac
+        append_delivery desktop "${stage}" sent "" "${measured_age}"
+        if IFS= read -r action && [ "${action}" = "default" ] && [ ! -e "${disarm_sentinel}" ]; then
+          XDG_RUNTIME_DIR="${runtime_dir}" DBUS_SESSION_BUS_ADDRESS="${bus_address}" \
+            timeout 5s xdg-open "${open_uri}" </dev/null >/dev/null 2>&1 || true
```
Comment (kriegcloud) [inline .claude/hooks/sequence-break-notifier.sh]: **P2 · Action listeners ignore bracket-resolved after wait closes**

Each `deliver_desktop` with `open_uri` backgrounds a `timeout 3600s` `notify-send --action` listener. On `action=default` the only suppress check is the global disarm sentinel — not `bracket_status`. Reminder/urgent skip new sends when resolved but do not cancel prior listeners, so up to three concurrent 1h listeners can still `xdg-open` after the wait already closed. Recheck bracket (or a per-request settle sentinel) before `xdg-open`, and cap to one live listener per session+target.

<details>
<summary>Prompt To Fix With AI</summary>

```markdown
This is a comment left during a code review.
Path: .claude/hooks/sequence-break-notifier.sh
Line: 450-452

Comment:
**Action listeners ignore bracket-resolved after wait closes**

Each `deliver_desktop` with `open_uri` backgrounds a `timeout 3600s` `notify-send --action` listener. On `action=default` the only suppress check is the global disarm sentinel — not `bracket_status`. Reminder/urgent skip new sends when resolved but do not cancel prior listeners, so up to three concurrent 1h listeners can still `xdg-open` after the wait already closed. Recheck bracket (or a per-request settle sentinel) before `xdg-open`, and cap to one live listener per session+target.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
```

</details>

File: .codex/hooks/hook-pulse.sh [hunk 6]
```diff
@@ -383,6 +383,20 @@ if [ "${notifier_rev}" != "log-only-0" ]; then
...
+      if [ "${TERM_PROGRAM:-}" = "ghostty" ]; then
+        if { exec 8>/dev/tty; } 2>/dev/null && [ -t 8 ]; then
+          notification_terminal="ghostty"
+        fi
+      fi
+      notification_uri="${BEEP_SEQUENCE_BREAK_OPEN_URI:-}"
+      if [ -z "${notification_uri}" ] &&
+        [ "${CODEX_INTERNAL_ORIGINATOR_OVERRIDE:-}" = "Codex Desktop" ] &&
+        [ -n "${CODEX_THREAD_ID:-}" ] && [ "${CODEX_THREAD_ID}" = "${raw_session_id}" ]; then
+        notification_uri="codex://threads/${raw_session_id}"
```
Comment (kriegcloud) [inline .codex/hooks/hook-pulse.sh:401]: **P2 · OPEN_URI bypasses Codex session-equality guard**

`BEEP_SEQUENCE_BREAK_OPEN_URI` is applied first; the `CODEX_THREAD_ID == raw_session_id` auto path runs only when that env is empty. The notifier allowlist then accepts any well-formed `codex://threads/<UUID>` with no equality check to the notifying session. Docs claim a child CLI cannot inherit a parent destination via `CODEX_THREAD_ID`; a standing exported `OPEN_URI` reopens that hole. For `codex-cli`, accept `OPEN_URI` only when it equals `codex://threads/${raw_session_id}` (same predicate as the auto path), otherwise clear before argv.

<details>
<summary>Prompt To Fix With AI</summary>

```markdown
This is a comment left during a code review.
Path: .codex/hooks/hook-pulse.sh
Line: 394-398

Comment:
**OPEN_URI bypasses Codex session-equality guard**

`BEEP_SEQUENCE_BREAK_OPEN_URI` is applied first; the `CODEX_THREAD_ID == raw_session_id` auto path runs only when that env is empty. The notifier allowlist then accepts any well-formed `codex://threads/<UUID>` with no equality check to the notifying session. Docs claim a child CLI cannot inherit a parent destination via `CODEX_THREAD_ID`; a standing exported `OPEN_URI` reopens that hole. For `codex-cli`, accept `OPEN_URI` only when it equals `codex://threads/${raw_session_id}` (same predicate as the auto path), otherwise clear before argv.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
```

</details>

File: .claude/hooks/sequence-break-notifier.sh [hunk 7]
```diff
@@ -385,6 +391,70 @@ notification_body() {
...
+  [ -n "${root}" ] || return 0
+  if command -v git >/dev/null 2>&1 && command -v timeout >/dev/null 2>&1; then
+    root="$(timeout 2s git -C "${origin_cwd}" rev-parse --show-toplevel 2>/dev/null)" || root="${origin_cwd%/}"
+    common="$(timeout 2s git -C "${origin_cwd}" rev-parse --path-format=absolute --git-common-dir 2>/dev/null)" || common=""
+  else
+    common=""
+  fi
+  case "${common}" in
+    */.git) clone="${common%/.git}" ;;
+    *) clone="${root}" ;;
```
Comment (kriegcloud) [inline .claude/hooks/sequence-break-notifier.sh:411]: **P2 · Bare-repo common-dir mislabels Clone/Worktree**

`desktop_origin` only treats `*/.git` as the owning clone path; anything else (including a bare common-dir like `/repos/beep.git`) falls through to `clone=root`. Linked worktrees against a bare repo then show `Clone: <worktree-basename>` with no Worktree segment, hiding the real bare repo name. Extend the case for `*.git` (and optional trailing `/.git/`) before comparing root vs clone.

<details>
<summary>Prompt To Fix With AI</summary>

```markdown
This is a comment left during a code review.
Path: .claude/hooks/sequence-break-notifier.sh
Line: 406-408

Comment:
**Bare-repo common-dir mislabels Clone/Worktree**

`desktop_origin` only treats `*/.git` as the owning clone path; anything else (including a bare common-dir like `/repos/beep.git`) falls through to `clone=root`. Linked worktrees against a bare repo then show `Clone: <worktree-basename>` with no Worktree segment, hiding the real bare repo name. Extend the case for `*.git` (and optional trailing `/.git/`) before comparing root vs clone.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
```

</details>

File: .claude/hooks/sequence-break-notifier.sh [hunk 8]
```diff
@@ -393,17 +463,46 @@ deliver_desktop() {
...
+  # Ghostty binds its native OSC notification to the emitting surface, including
+  # the tab/split click action. FD 8 was opened before the hook detached. Never
+  # write terminal escapes to hook stdout or reopen a potentially recycled PTY.
+  if [ -z "${open_uri}" ] && [ "${origin_terminal}" = "ghostty" ] && [ -t 8 ]; then
+    local osc_body
+    # Ghostty uses the body as its notification ID. Distinguish simultaneous
+    # sessions in the same checkout so one cannot replace the other's action.
+    osc_body="$(jq -nr --arg body "${body} [${session_id:0:12}]" '$body | gsub("[\u0000-\u001f\u007f-\u009f;]"; " ")')"
+    if [ ! -e "${disarm_sentinel}" ] &&
+      timeout 2s bash -c 'printf "\\033]777;notify;%s;%s\\033\\\\" "$1" "$2" >&8' bash "${title}" "${osc_body}"; then
```
Comment (kriegcloud) [inline .claude/hooks/sequence-break-notifier.sh]: **P2 · Isolate Ghostty FD 8 from ntfy bearer reuse**

This new Ghostty path requires `[ -t 8 ]` on the inherited tty, but `deliver_stage` still runs `deliver_ntfy` afterward. When a bearer token is set, `deliver_ntfy` does `exec 8< <(Authorization…)` then `exec 8<&-`, closing FD 8 so reminder/urgent re-entries fail the Ghostty gate and silently fall back to plain `notify-send`. Keep the Ghostty tty on a dedicated FD for the worker lifetime; put the ntfy bearer on a different FD.

<details>
<summary>Prompt To Fix With AI</summary>

```markdown
This is a comment left during a code review.
Path: .claude/hooks/sequence-break-notifier.sh
Line: 487-496

Comment:
**Isolate Ghostty FD 8 from ntfy bearer reuse**

This new Ghostty path requires `[ -t 8 ]` on the inherited tty, but `deliver_stage` still runs `deliver_ntfy` afterward. When a bearer token is set, `deliver_ntfy` does `exec 8< <(Authorization…)` then `exec 8<&-`, closing FD 8 so reminder/urgent re-entries fail the Ghostty gate and silently fall back to plain `notify-send`. Keep the Ghostty tty on a dedicated FD for the worker lifetime; put the ntfy bearer on a different FD.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
```

</details>

**Review by kriegcloud — COMMENTED** [review]
Greptile-style review (r1): four verified P2 findings on notification-origin presentation/lifecycle. No P0/P1. COMMENT only — mergeable_state blocked; do not APPROVE.

**Comment by kriegcloud** [comment]
## Confidence Score: 4/5

Safe to iterate on presentation/lifecycle P2s; do not merge while checks keep mergeable_state blocked — no P0/P1 found.

## Findings

1. **P2 · Isolate Ghostty FD 8 from ntfy bearer reuse** → `.claude/hooks/sequence-break-notifier.sh:487-496`
2. **P2 · Action listeners ignore bracket-resolved after wait closes** → `.claude/hooks/sequence-break-notifier.sh:450-452`
3. **P2 · OPEN_URI bypasses Codex session-equality guard** → `.codex/hooks/hook-pulse.sh:394-398`
4. **P2 · Bare-repo common-dir mislabels Clone/Worktree** → `.claude/hooks/sequence-break-notifier.sh:406-408`

<details>
<summary>Fix with agent prompt</summary>

### Issue 1
path: .claude/hooks/sequence-break-notifier.sh:487-496
Ghostty OSC requires `[ -t 8 ]`, but `deliver_ntfy` reuses FD 8 for the bearer header and closes it. Reserve a dedicated FD for the Ghostty tty for the worker lifetime; put the ntfy bearer on a different FD so reminder/urgent stages can still satisfy the Ghostty gate.

### Issue 2
path: .claude/hooks/sequence-break-notifier.sh:450-452
Before `xdg-open` on `action=default`, re-run `bracket_status` (or honor a per-request settle sentinel). Cap to one live 3600s action listener per session+target; cancel/replace prior listeners when reminder/urgent fire or when the wait resolves. Add a test that prints default after resolved/disarm and asserts no `xdg-open`.

### Issue 3
path: .codex/hooks/hook-pulse.sh:394-398
For `codex-cli`, after reading `BEEP_SEQUENCE_BREAK_OPEN_URI`, accept only when it equals `codex://threads/${raw_session_id}` (same predicate as the auto path); otherwise clear before argv. Document that OPEN_URI must not be a process-tree-wide export across concurrent sessions.

### Issue 4
path: .claude/hooks/sequence-break-notifier.sh:406-408
Extend the `desktop_origin` common-dir case for bare `*.git` (and optional trailing `/.git/`): strip `/.git` when present; for bare `*.git` use that path as clone; then compare root vs clone for the Worktree segment.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
</details>

<details open>
<summary><h3>Summary</h3></summary>

- What PR does: Adds desktop-only notification origin labels (agent, clone/worktree), Ghostty OSC surface return via inherited FD 8, and allowlisted Open-task URIs for Codex/Claude Desktop, keeping origin/URI out of evidence and ntfy.
- Improvements: Content-free ledger/privacy invariants hold; URI allowlist rejects arbitrary/https/cross-agent routes; disarm still gates worker entry and actions; regression tests cover labels, worktree split, action open, and reject list.
- Remaining issues: FD 8 conflict with ntfy bearer closes Ghostty for later stages; stacked 1h action listeners ignore bracket-resolved; `BEEP_SEQUENCE_BREAK_OPEN_URI` bypasses the Codex session-equality guard claimed in docs; bare-repo common-dir mislabels Clone/Worktree. Summary-only: Desktop/Ghostty body suffix can outlive fallthrough delivery; Junior doc/lexicon nits (OPEN_URI argv contract, `notification_title` ntfy-only, OSC vs `@html`, argv table, Codex/ChatGPT naming).
- Risk tier L1 · seats: Domain Modeler, Reliability Adversary, Junior · deep_meaning_used:false (DankStation offline; structural graft unavailable on box)
- Blast radius (light): Claude/Codex hook-pulse launch seam → shared `.claude/hooks/sequence-break-notifier.sh` → desktop notify-send/OSC/action + ntfy; permission decision path and schema-owned ledger fields unchanged.

</details>

<details open>
<summary><h3>Diagram</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
  Hook[hook-pulse PermissionRequest] --> Args[cwd uri terminal args]
  Args --> Notifier[sequence-break-notifier]
  Notifier --> Origin[desktop_origin label]
  Notifier --> Allow[URI allowlist]
  Allow -->|valid| Action[notify-send action + xdg-open]
  Allow -->|empty + ghostty FD8| OSC[OSC 777]
  Notifier --> Ntfy[ntfy generic title/body]
  Ntfy -->|token uses FD8| Conflict[FD8 tty lost for later stages]
```

</details>

<sub>Reviews (1) · Last reviewed commit: ["feat(tooling): identify agent notification origins"](https://github.com/beep-effect/beep-effect/commit/544cdb4281e313e3cf01569579be60d6f3d3bf66)</sub>

File: .claude/hooks/sequence-break-notifier.sh (hunk 2 above)
Comment (kriegcloud) [reply .claude/hooks/sequence-break-notifier.sh:459]: Fixed in 180af8f679. The worker now retains the background listener PID and damps later desktop stages while the persistent notification is still actionable. This bounds each wait to one live listener, with the existing one-hour deadline. Phone escalation continues. The new regression exercises initial/reminder/urgent and asserts one listener invocation and two damped desktop receipts; the runbook documents the behavior.

File: packages/tooling/library/ai-metrics/test/sequence-break.test.ts (hunk 3 above)
Comment (kriegcloud) [reply packages/tooling/library/ai-metrics/test/sequence-break.test.ts]: Fixed in 180af8f679. runNotifier now forwards originTerminal and can run through a real util-linux script PTY with the reserved terminal descriptor. The regression verifies three OSC 777 notifications, no notify-send fallback, and three authenticated ntfy sends across initial/reminder/urgent. All 87 focused tests pass.

File: .claude/hooks/sequence-break-notifier.sh (hunk 4 above)
Comment (kriegcloud) [reply .claude/hooks/sequence-break-notifier.sh:503]: Fixed in 180af8f679. Both hooks now preserve Ghostty on FD 7, and the worker emits on FD 7. ntfy retains FD 8. A real PTY regression asserts all three Ghostty stages still emit with authenticated ntfy between them, and verifies all six delivery receipts are sent. The separate detached-hook smoke test also passes.

File: .claude/hooks/sequence-break-notifier.sh (hunk 5 above)
Comment (kriegcloud) [reply .claude/hooks/sequence-break-notifier.sh]: Fixed in 180af8f679. The click callback now rechecks bracket_status and opens only while the original wait remains open, in addition to the disarm check. Later desktop stages reuse the existing actionable notification rather than spawning more listeners. A smoke test closed the bracket before emitting the default action and confirmed no xdg-open call; the normal exact-session action test continues to pass.

File: .codex/hooks/hook-pulse.sh (hunk 6 above)
Comment (kriegcloud) [reply .codex/hooks/hook-pulse.sh:401]: Fixed in 180af8f679. The Codex hook now clears any explicit override that does not exactly equal codex://threads/${raw_session_id} before considering the automatic Desktop mapping. A smoke test supplied an inherited well-formed URI for another session and confirmed no action was offered or app opened. The matching automatic route still opens successfully.

File: .claude/hooks/sequence-break-notifier.sh (hunk 7 above)
Comment (kriegcloud) [reply .claude/hooks/sequence-break-notifier.sh:411]: Fixed in 180af8f679. The common directory is normalized for a trailing slash and a bare *.git repository is retained as the owning clone. A regression using /workspace/owning.git/ with a separate worktree now asserts Clone: owning.git and Worktree: feature-lane.

File: .claude/hooks/sequence-break-notifier.sh (hunk 8 above)
Comment (kriegcloud) [reply .claude/hooks/sequence-break-notifier.sh]: Fixed in 180af8f679. Ghostty now uses dedicated FD 7 in both hook writers and the worker, leaving ntfy bearer reads on FD 8. The new PTY regression proves initial/reminder/urgent OSC delivery survives all three authenticated phone sends.

**Review by kriegcloud — COMMENTED** [review]
### Re-review (head `180af8f`)

Prior Greptile findings are addressed in [180af8f679](https://github.com/beep-effect/beep-effect/commit/180af8f6793a4b77aeace2d3851089b73536bb2d). Verified: FD7/FD8 isolation, single-live listener + `bracket_status` before `xdg-open`, Codex `OPEN_URI` session equality, bare-repo labeling, Ghostty PTY + ntfy regression, multi-stage damping.

**No new inline findings.** Confidence 4/5. `mergeable_state: blocked` — COMMENT only (no approve).

**Comment by kriegcloud** [comment]
## Confidence Score: 4/5

Re-review clean on `180af8f` — prior findings are addressed; minor polish only (merge still blocked by required checks).

<details open>
<summary><h3>Summary</h3></summary>

- **What:** Local desktop wait notifications now show agent / clone / worktree, with Ghostty OSC surface targeting and allowlisted Codex/Claude Desktop Open-task actions. Origin and navigation stay out of evidence + ntfy.
- **Harden pass (`180af8f`):** Verified closed vs prior threads — Ghostty on **FD7** isolated from ntfy bearer **FD8**; single live action listener with desktop stage damping; `bracket_status=open` + disarm before `xdg-open`; Codex `OPEN_URI` must equal `codex://threads/${raw_session_id}`; bare `*.git` common-dir labeling; `runNotifier` forwards `originTerminal` with PTY Ghostty+ntfy regression; multi-stage listener test.
- **Remaining:** No new actionable P0/P1/P2. Residual notes (summary only): Claude `OPEN_URI` stays launcher-owned (no CLI↔Desktop id equality); ledger reason `storm-damped` now also means “live Open-task listener”; PID `kill -0` liveness across reminder sleep is acceptable for this pass.
- **Risk:** L1 · seats Domain Modeler, Reliability Adversary, Junior · Category Theorist skipped (shell-dominant, no Effect/Schema laws) · `deep_meaning_used: false` (DankStation offline; structural graft unavailable).
- This comment **supersedes** prior greptile_style_summary ids [5691741919](https://github.com/beep-effect/beep-effect/pull/1144#issuecomment-5691741919) and [5691724190](https://github.com/beep-effect/beep-effect/pull/1144#issuecomment-5691724190).

</details>

<details open>
<summary><h3>Diagram</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
  HP[hook-pulse PermissionRequest] --> FD7{Ghostty? exec 7>/dev/tty}
  HP --> URI[OPEN_URI / Codex Desktop bind]
  FD7 --> N[sequence-break-notifier]
  URI --> N
  N --> Allow{URI allowlist}
  Allow -->|ok + xdg-open| L[single action listener]
  Allow -->|empty + FD7 tty| G[OSC 777 on FD7]
  Allow -->|reject| P[plain notify-send]
  L -->|click| B{bracket open + !disarm}
  B -->|yes| X[xdg-open session URI]
  B -->|no| Z[no-op]
  N --> NTFY[deliver_ntfy FD8 bearer]
  L -.->|listener live| Damp[later desktop storm-damped]
  Damp --> NTFY
```

</details>

<sub>Reviews (3) · Last reviewed commit: ["fix(tooling): harden notification routing and listener lifecycle"](https://github.com/beep-effect/beep-effect/commit/180af8f6793a4b77aeace2d3851089b73536bb2d)</sub>
